### PR TITLE
[AMD][DSV4] DSV4 MTP graph + sparse triton attn optimizations

### PR DESCRIPTION
**Pull request by kkHuang-amd**
Co-authored-by: @1am9trash @RolaoDenthu @yichiche @hubertlu-tw @sogalin @HaiShaw @yctseng0211 @bingxche @amd-danli103 @Raiden-Makoto

## Motivation

The long-standing draft #25552 collected 85 commits of AMD / DSV4 work over several rebases. Most of the optimization PRs it tracked (compressor opts, fused topk, sparse MLA kernels, fused compress, etc.) have already been merged into `main` through smaller follow-up PRs (e.g. #26208, #25251, #25878, #25977, #26014). What remains is a small but meaningful set of changes needed to:

1. Make DSV4 + EAGLE/MTP speculative decoding work correctly under CUDA-graph capture on ROCm. The current HIP radix backend mis-handles per-step draft `out_cache_loc` slicing and trips planner invariants inside `init_forward_metadata_target_verify` when `SGLANG_PREP_IN_CUDA_GRAPH` is on, leading to graph-capture failure / wrong KV slots being written during target verification.
2. Land the last round of ROCm performance wins for DSV4 decode/extend that did not make it into the previously merged sub-PRs:
   - Fused Triton `mhc_post_pre` for low concurrency (enabled behind `SGLANG_OPT_USE_TRITON_FUSED_MHC`).
   - `aiter.greedy_sample` fast path for all-greedy sampling on ROCm.
   - Enable `w_qkv` fusion on ROCm and tidy up the HIP multi-stream path.
3. Provide a clean, single-commit, conflict-free landing of the above so reviewers do not have to wade through 85 commits / merge commits.

The diff against `main` is small (10 files, +656 / -65), self-contained to ROCm-specific code paths plus a single shared `deepseek_common/amd/` helper, and does not change CUDA behavior outside of `is_hip()` guarded branches.

## Modifications

- `python/sglang/srt/environ.py`
  - Add `SGLANG_OPT_USE_TRITON_FUSED_MHC` (default `True`) to gate the new fused mHC post+pre helper.

- `python/sglang/srt/layers/sampler.py`
  - When `SGLANG_USE_AITER` is set on ROCm, replace the `torch.argmax` all-greedy fallback with `aiter.greedy_sample` writing into a preallocated `int32` buffer. Non-HIP / non-aiter paths are unchanged.

- `python/sglang/srt/speculative/draft_utils.py`
  - In `DraftBackendFactory._create_dsv4_decode_backend` and `_create_dsv4_prefill_backend`, dispatch to `deepseek_v4_backend_hip_radix.DeepseekV4MultiStepBackend` / `DeepseekV4HipRadixBackend` on ROCm and fall back to the existing CUDA `deepseek_v4_backend` elsewhere. This is the actual bug-fix for "DSV4 MTP graph backend handling on HIP".

- `python/sglang/srt/layers/attention/deepseek_v4_backend_hip_radix.py`
  - `init_forward_metadata_target_verify`: always build metadata eagerly via `init_forward_metadata_target_verify_old` for the HIP path. The previous `SGLANG_PREP_IN_CUDA_GRAPH` raw/lazy-upgrade branch hit DSV4+EAGLE planner invariants during graph capture.
  - `init_forward_metadata_target_verify_old`: normalize `extend_seq_lens` (allocate per-batch when missing, clamp to `seq_lens`, floor at 1) so target-verify works for all draft batch shapes.
  - `init_forward_metadata`: in decode/idle mode slice the shared multi-step `out_cache_loc` with `per_step_draft_out_cache_loc(... )[self.speculative_step_id]` before passing it into metadata, matching how DSV4 bakes the c4/c128 KV write target into metadata. Also forward `forward_batch.extend_seq_lens` into target-verify metadata.

- `python/sglang/srt/layers/attention/nsa/triton_decode/triton_mla_kernels_decode_fused.py`, `triton_mla_kernels_decode_optimized.py`
  - Wire the fused / optimized Triton sparse MLA decode kernels into the new "single fused gather + attention" prefill dispatch path used by HIP. These are the kernel-side updates for the prefill-fusion item.

- `python/sglang/srt/models/deepseek_common/amd/__init__.py` (new)
- `python/sglang/srt/models/deepseek_common/amd/deepseek_v4_fused_mhc.py` (new, 158 lines)
  - New shared AMD helper `try_fused_hc_post_pre(...)` that performs the fused `hc_post + hc_pre` (matmul + scale/base + sinkhorn + RMSNorm) and reports back whether the post-attention layernorm was fused. Returns `None` to signal the caller to fall back to the unfused path.

- `python/sglang/srt/models/deepseek_v4.py`
  - Add `_freqs_cis_to_cos_sin` cache so all layers sharing a `precompute_freqs_cis` result reuse the same `(cos, sin)` bf16 contiguous tables.
  - In `MQALayer`: enable `fuse_wqa_wkv` on ROCm (drop the `not _is_hip` gate) and pass `attn_backend` into the compressor call when fused. Guard the multi-stream alt-stream creation to avoid HIP cases where `self.compressor is None`.
  - In `DeepseekV4DecoderLayer.forward`: try the fused `try_fused_hc_post_pre(...)` first; consume its `(residual, hidden_states, post, comb, norm_fused)` tuple and only run the unfused `hc_post` / `hc_pre` / `post_attention_layernorm` path when the helper returns `None`. This preserves the existing `norm_fused` return contract.
  - In `DeepseekV4ForCausalLM.load_weights`: also enable `fuse_wqa_wkv` weight bucketing on ROCm.

- `test/registered/ops/test_aiter_greedy_sample_amd.py` (new, 289 lines)
  - Registered unit test covering the `aiter.greedy_sample` sampler integration on AMD, exercising small / large vocab, tied logits, and parity vs `torch.argmax` across dtypes and batch sizes.

## AMD Accuracy Tests
<details><summary>server command</summary>
<p>

```
export SGLANG_OPT_USE_JIT_KERNEL_FUSED_TOPK=true

export SGLANG_OPT_DEEPGEMM_HC_PRENORM=false

export SGLANG_ENABLE_THINKING=1
export SGLANG_USE_AITER=1
export SGLANG_USE_ROCM700A=1
export SGLANG_FP8_PAGED_MQA_LOGITS_TORCH=1

export SGLANG_OPT_USE_OVERLAP_STORE_CACHE=false
export SGLANG_OPT_USE_FUSED_STORE_CACHE=true

# changed
export SGLANG_OPT_USE_FUSED_COMPRESS=true
export SGLANG_TOPK_TRANSFORM_512_TORCH=0
export SGLANG_OPT_USE_TILELANG_INDEXER=true
export SGLANG_OPT_USE_AITER_INDEXER=false
export SGLANG_HACK_FLASHMLA_BACKEND=triton
export SGLANG_REASONING_EFFORT=max
export SGLANG_OPT_USE_AITER_MHC_PRE=true
export SGLANG_OPT_USE_AITER_MHC_POST=true
export SGLANG_OPT_USE_FUSED_HASH_TOPK=true
export SGLANG_OPT_FUSE_WQA_WKV=true

export SGLANG_OPT_USE_FUSED_QK_NORM_ROPE=true
export SGLANG_OPT_USE_FUSED_CLAMP_ACT_MUL=true
export SGLANG_OPT_USE_FUSED_COMPRESS_TRITON=true

export SGLANG_OPT_FP8_WO_A_GEMM=false
export SGLANG_OPT_USE_TOPK_V2=false

export SGLANG_OPT_USE_COMPRESSOR_V2=false
export SGLANG_OPT_USE_JIT_NORM=false

export SGLANG_OPT_DPSK_V4_RADIX=1
export SGLANG_OPT_C4_SPARSE_TOPK=512

export SGLANG_OPT_USE_JIT_INDEXER_METADATA=false
export SGLANG_OPT_USE_MULTI_STREAM_OVERLAP=false

export AITER_BF16_FP8_MOE_BOUND=1

MODEL=/dockerx/data/models/DeepSeek-V4-Pro/

python3 -m sglang.launch_server \
    --model-path ${MODEL} \
    --trust-remote-code \
    --tp 8 \
    --dp 8 --enable-dp-attention --enable-prefill-delayer \
    --disable-radix-cache \
    --attention-backend compressed \
    --max-running-request 256 \
    --page-size 256 \
    --chunked-prefill-size 8192 \
    --port 8000 \
    --speculative-algorithm EAGLE \
    --speculative-num-steps 2 \
    --speculative-num-draft-tokens 3 \
    --speculative-eagle-topk 1 \
    --disable-shared-experts-fusion \
    --tool-call-parser deepseekv4 \
    --reasoning-parser deepseek-v4
```

</p>
</details> 

<details><summary>client command</summary>
<p>

```
python3 benchmark/gsm8k/bench_sglang.py --num-questions 1319 --parallel 1319 --num-shots 5 --port 8000
```

</p>
</details> 

**Result**

```
/sgl-workspace/sglang-cursor# python3 benchmark/gsm8k/bench_sglang.py --num-questions 1319 --parallel 1319 --num-shots 5 --port 8000
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [01:13<00:00, 17.97it/s]
Accuracy: 0.950
Invalid: 0.000
Latency: 73.386 s
Output throughput: 1636.519 token/s
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #26500636036](https://github.com/sgl-project/sglang/actions/runs/26500636036)
Latest PR Test (Extra): :x: [Run #26500636297](https://github.com/sgl-project/sglang/actions/runs/26500636297)

### CODE REVIEW

**Comment by yctseng0211** [comment]
AMD failures are all known, irrelevant to this PR : 
image
WIP PRs
https://github.com/sgl-project/sglang/pull/26442, https://github.com/sgl-project/sglang/pull/26243 for [PR Test (AMD) / stage-b-test-1-gpu-small-amd (linux-mi325-1gpu-sglang, 7)](https://github.com/sgl-project/sglang/actions/runs/26493678169/job/78018238231?pr=26383) and [PR Test (AMD) / stage-b-test-1-gpu-small-amd-nondeterministic (linux-mi325-1gpu-sglang)](https://github.com/sgl-project/sglang/actions/runs/26493678169/job/78018238059?pr=26383)
https://github.com/sgl-project/sglang/pull/26178 for [PR Test (AMD) / stage-b-test-1-gpu-large-amd (linux-mi325-1gpu-sglang, 1) (pull_request)](https://github.com/sgl-project/sglang/actions/runs/26493678169/job/78018238090?pr=26383)

the skipped tests are also verified, the failures are all known.
https://github.com/sgl-project/sglang/actions/runs/26453711426

**Review by yctseng0211 — APPROVED** [review]
LGTM

**Review by 1am9trash — APPROVED** [review]

**Comment by HaiShaw** [comment]
@yctseng0211 please confirm NV test results when all done.

**Review by HaiShaw — APPROVED** [review]

**Comment by HaiShaw** [comment]
@amd-bot ci-status
